src/postprocessor/compress.rs

```diff
@@ -92,39 +153,61 @@ impl Postprocessor for Lz4 {
     fn process(&mut self, _ingres_ns: u64, _egress_ns: u64, data: &[u8]) -> Result<Vec<Vec<u8>>> {
         use lz4::EncoderBuilder;
         let buffer = Vec::<u8>::new();
-        let mut encoder = EncoderBuilder::new().level(4).build(buffer)?;
+        let mut encoder = EncoderBuilder::new().level(self.compression_level).build(buffer)?;
         encoder.write_all(data)?;
         Ok(vec![encoder.finish().0])
     }
 }
+impl Default for Lz4 {
+   fn default() -> Self {
+        Self {
+            compression_level:4,
+        }
+   } 
+}
 
-#[derive(Clone, Default, Debug)]
-struct Zstd {}
+#[derive(Clone, Debug)]
+struct Zstd {
+    compression_level: i32,
+}
+impl Zstd {
+    pub fn with_config(level: i32) -> Self {
+        Self {
+            compression_level: level,
+        }
+    }
+}
 impl Postprocessor for Zstd {
     fn name(&self) -> &str {
         "zstd"
     }
 
     fn process(&mut self, _ingres_ns: u64, _egress_ns: u64, data: &[u8]) -> Result<Vec<Vec<u8>>> {
         // Value of 0 indicates default level for encode.
-        let compressed = zstd::encode_all(data, 0)?;
+        let compressed = zstd::encode_all(data, self.compression_level)?;
         Ok(vec![compressed])
     }
 }
-
+impl Default for Zstd {
+   fn default() -> Self {
+        Self {
+            compression_level:0,
+        }
+   } 
+}
 pub(crate) struct Compress {
     codec: Box<dyn Postprocessor>,
 }
 impl Compress {
     pub(crate) fn from_config(config: Option<&Value>) -> Result<Self> {
         let codec: Box<dyn Postprocessor> =
             match config.get_str("algorithm").ok_or("Missing algorithm")? {
-                "gzip" => Box::new(Gzip::default()),
-                "zlib" => Box::new(Zlib::default()),
-                "xz2" => Box::new(Xz2::default()),
-                "snappy" => Box::new(Snappy::default()),
-                "lz4" => Box::new(Lz4::default()),
-                "zstd" => Box::new(Zstd::default()),
+                "gzip" => Algorithm::postprocessor(config, Algorithm::Gzip),
```

Comment:
idea: What about implementing https://doc.rust-lang.org/std/str/trait.FromStr.html for `Algorithm`, use it here and then maybe implement a method like `fn into_postprocessor(self, config: Option<&Value>) -> Result<Box<dyn Postprocessor>>` and then use it here as well.

Reply:
I like the idea of having a Algoritm enum it's neat, it's much cleaner than the string matching we had before! So Matthias has a point `FromStr` would be neat here or `Deserialize` and a CompressorConfig to combine algorithm and level, could be an option if levels exist for all compression algorithms?

Reply:
I suppose so, but what about algorithms like "snappy" which do not have levels?

Reply:
That's a good point then `Deserialize` won't work too well.